sys/net/gnrc/network_layer/ipv6/nib/_nib-internal.c

```diff
@@ -0,0 +1,387 @@
+    _nib_onl_entry_t *first = (_nib_onl_entry_t *)clist_lpop(&_next_removable);
+    _nib_onl_entry_t *tmp = first, *res = NULL;
+
+    DEBUG("nib: Searching for replaceable entries (addr = %s, iface = %u)\n",
+          ipv6_addr_to_str(addr_str, addr, sizeof(addr_str)), iface);
+    if (tmp != NULL) {
+        do {
+            if (_is_gc(tmp)) {
+                DEBUG("nib: Removing neighbor cache entry (addr = %s, "
+                      "iface = %u) ",
+                      ipv6_addr_to_str(addr_str, &tmp->ipv6,
+                                       sizeof(addr_str)),
+                      _nib_onl_get_if(tmp));
+                DEBUG("for (addr = %s, iface = %u)\n",
+                      ipv6_addr_to_str(addr_str, addr, sizeof(addr_str)),
+                      iface);
+                res = tmp;
+                res->mode = _EMPTY;
+                _override_node(addr, iface, res);
+                /* masked above already */
+                res->info = cstate;
+                res->mode = _NC;
+            }
+            else {
+                /* requeue if not garbage collectible at the moment */
+                DEBUG("nib: Requeing (addr = %s, iface = %u)\n",
+                      ipv6_addr_to_str(addr_str, &tmp->ipv6,
+                                       sizeof(addr_str)),
+                      _nib_onl_get_if(tmp));
+                clist_rpush(&_next_removable, (clist_node_t *)tmp);
+                tmp = (_nib_onl_entry_t *)clist_lpop(&_next_removable);
+            }
+        } while (tmp != first);
+    }
+    return res;
+}
+
+_nib_onl_entry_t *_nib_nc_add(const ipv6_addr_t *addr, unsigned iface,
+                              uint16_t cstate)
+{
+    cstate &= GNRC_IPV6_NIB_NC_INFO_NUD_STATE_MASK;
+    assert(cstate != GNRC_IPV6_NIB_NC_INFO_NUD_STATE_DELAY);
+    assert(cstate != GNRC_IPV6_NIB_NC_INFO_NUD_STATE_PROBE);
+    assert(cstate != GNRC_IPV6_NIB_NC_INFO_NUD_STATE_REACHABLE);
+    _nib_onl_entry_t *node = _nib_onl_alloc(addr, iface);
+    if (node != NULL) {
+        DEBUG("nib: Adding to neighbor cache (addr = %s, iface = %u)\n",
+              ipv6_addr_to_str(addr_str, addr, sizeof(addr_str)), iface);
+        if (!(node->mode & _NC)) {
+            node->info &= ~GNRC_IPV6_NIB_NC_INFO_NUD_STATE_MASK;
+            /* masked above already */
+            node->info |= cstate;
+            node->mode |= _NC;
+        }
+        if (node->next == NULL) {
+            DEBUG("nib: queueing (addr = %s, iface = %u) for potential removal\n",
+                  ipv6_addr_to_str(addr_str, addr, sizeof(addr_str)), iface);
+            /* add to next removable list, if not already in it */
+            clist_rpush(&_next_removable, (clist_node_t *)node);
+        }
+    }
+    else {
+        node = _cache_out_onl_entry(addr, iface, cstate);
+    }
+    return node;
+}
+
+_nib_onl_entry_t *_nib_onl_iter(const _nib_onl_entry_t *last)
+{
+    for (const _nib_onl_entry_t *node = (last) ? last + 1 : _nodes;
```

Comment:
I don't understand the motivation for `const` here, since the array that is iterated over is also not an array of `const` elements. IMO there's no need for that and makes the below lines less verbose.